net/gnrc/netif: Added CC1xxx adaption layer

maribu opened the pull request:
### Contribution description
This adaption layer is intended to be used by a rewrite of the CC110x driver and a new CC1200 driver that is currently being developed for RIOT. It does not have any users yet.

### Testing procedure
Not possible now, to be honest. But I will adapt PR https://github.com/RIOT-OS/RIOT/pull/10340 to this PR and rebase it on top of this, if this approach is accepted.

### Issues/PRs references
Intended users of this PR:
 - https://github.com/RIOT-OS/RIOT/pull/10340
 - https://github.com/RIOT-OS/RIOT/pull/7598

maribu commented:
@miri64: I just named it `cc1xxx` for now and did not generalize the layer 2 broadcast. Maybe it is better to not over-engineer things up front when it is still unclear if there will ever be other users this adaption layer than `cc110x` and `cc1200`

maribu commented:
* Fixed naming of include guard and squashed. (No one started to review now, I hope.)

maribu commented:
Here is way to test it: https://github.com/maribu/RIOT/commits/cc110x_rewrite2
This branch contains:
 - This PR (but rebased against current master)
 - PR https://github.com/RIOT-OS/RIOT/pull/10340 adapted to this PR and squashed

miri64 commented on sys/net/gnrc/netif/cc1xxx/gnrc_netif_cc1xxx.c:
```diff
@@ -0,0 +1,172 @@
+/*
```
Mhh, following https://github.com/RIOT-OS/RIOT/pull/9917#issuecomment-450176314 I'm now not sure `sys/net/gnrc/netif/cc1xxx` is the right place for this. Maybe `drivers/cc1xxx_common`? This would also allow for some "cross polination" of code for the two `cc1xxx` drivers, if some code duplication appears.

maribu replied:
> Maybe `drivers/cc1xxx_common`?

Honestly, I do not have the deep understanding about RIOT's structure to have a profound opinion for or against moving it. But `drivers/cc1xxx_common` sounds good to me, so I will move it there.

> This would also allow for some "cross polination" of code for the two cc1xxx drivers, if some code duplication appears.

This would be nice indeed and is good argument for `drivers/cc1xxx_common` :-)

maribu replied:
Done. I will rebase https://github.com/maribu/RIOT/commits/cc110x_rewrite2 on top of this to test it

miri64 commented:
> `fixup! Removed file not belonging to this PR`

Please don't break the old driver before the new one is merged ;-)

miri64 commented:
The CI recognizes that the PR needs squashing due to your commit messages ;-). Setting the label unnecessary, might require unnecessary re-builds later on.

maribu commented on drivers/cc1xxx_common/gnrc_netif_cc1xxx.c:
```diff
@@ -0,0 +1,172 @@
+        /* check that destination address is valid */
+        assert(netif_hdr->dst_l2addr_len > 0);
+        uint8_t *addr = gnrc_netif_hdr_get_dst_addr(netif_hdr);
+        l2hdr.dest_addr = addr[netif_hdr->dst_l2addr_len - 1];
```
Cargo cult programming I guess:

https://github.com/RIOT-OS/RIOT/blob/d99d72c583789bba3d8696274358af2d52d19ac9/drivers/cc110x/gnrc_cc110x/gnrc_cc110x.c#L60-L61

Btw: Would it not be better to even use

```
assert(netif_hdr->dst_l2addr_len == 0);
uint8_t *addr = gnrc_netif_hdr_get_dst_addr(netif_hdr);
l2hdr.dest_addr = addr[0];
```
(Note the change assert.) For CC1xxx devices Layer 2 addresses will always be 1 byte long

miri64 replied:
> `assert(netif_hdr->dst_l2addr_len == 0);`

Should this be `== 1` then?

miri64 replied:
I prefer to not check the exact length of the address however. At least I remember, that we once stated, that addresses could be whatever length in input and the driver just takes the most significant bytes.

maribu commented on drivers/include/cc1xxx_common.h:
```diff
@@ -0,0 +1,139 @@
+ * ======================
+ *
+ * This adaption layer is written to be used by CC110x and CC1200 transceivers,
+ * but any transceiver using that should transfer 6LoWPAN and uses layer 2
```
Any payload fitting into the MTU of 255 Bytes minus the header (2 bytes currently). However, currently layer 2 does not have a field to identify the protocol of the frames payload, which makes using multiple protocols simultaneously difficult. I plan to extend the layer 2 format in later PRs to address this.

Still, the this code would allow even yet to overwrite `C1XXX_DEFAULT_PROTOCOL` to use carry other protocols. So indeed, that commend is misleading.

miri64 replied:
> However, currently layer 2 does not have a field to identify the protocol of the frames payload, which makes using multiple protocols simultaneously difficult. I plan to extend the layer 2 format in later PRs to address this.
>
> Still, the this code would allow even yet to overwrite C1XXX_DEFAULT_PROTOCOL to use carry other protocols. So indeed, that commend is misleading.

Yepp, especially if you try using GNRC with out the 6LoWPAN module with this driver ;-) (which AFAIK `examples/default` does.

maribu commented on drivers/include/cc1xxx_common.h:
```diff
@@ -0,0 +1,139 @@
+ * This adaption layer assumes that the layer 2 address `0x00` is reserved for
+ * layer 2 broadcast, which is true for CC110x and CC1200 transceivers (provided
+ * they are configured accordingly). If more users of this adaption layers are
+ * added, this behaviour might needs to be more generalized.
```
+1 on not using a magic number.

But allowing it to be overwritten can cause trouble. Assuming another user of this adaption layer pops up that uses the code unmodified but overwrites the broadcast address e.g. to `0xff`, this device could not be used in conjunction with CC110x or CC1200 transceivers. If this case appears, I would favor to e.g. extend `cc1xxx_t` e.g. to

```C
typedef struct {
    netdev_t netdev;        /**< RIOT's interface to this driver */
    uint8_t addr;           /**< Layer 2 address of this device */
    uint8_t bcast_addr;     /**< Layer 2 address used in broadcasts  */
} cc1xxx_t;
```

(But for now I would save the one byte, as it is not yet required.)

miri64 replied:
I know, that simultaneous usage of both wouldn't be possible then. Since this is a more hypothetical case however, I could live with that.

> (But for now I would save the one byte, as it is not yet required.)

But only on certain 8-bit platforms (however, I believe all 8-bit platforms we support align to 16-bit). However, due to the required initialization this would require some more ROM.

miri64 replied:
The define resolves the conversation for me for now. If similar link-layers with different broadcast addresses arise we can change that.

maribu replied:
> If similar link-layers with different broadcast addresses arise we can change that.

Maybe by then a solution how to use the same code on simultaneously is obvious.

maribu commented on drivers/include/cc1xxx_common.h:
```diff
@@ -0,0 +1,139 @@
+ * This includes only the Destination and Source address, as only those have
+ * impact on the Length Filed as expected/set by the transceiver
+ */
+#define CC1XXX_HEADER_SIZE              2
```
Indeed, this no longer makes sense. (Prior versions of `cc1xxx_l2hdr_t` also included the length field, so `sizeof(cc1xxx_l2hdr_t)` was previously one byte greater than the header length for the length field's point of view.)

miri64 commented on drivers/cc1xxx_common/gnrc_netif_cc1xxx.c:
```diff
@@ -0,0 +1,171 @@
+    if (payload == NULL) {
+        DEBUG("[cc1xxx-gnrc] recv: unable to allocate space in the pktbuf\n");
+        /* tell the driver to drop the packet */
+        dev->driver->recv(dev, NULL, 1, NULL);
```
Nitpicky: magic number. I propose to use `pktlen` here. This might even be somewhat of an optimization, since `pktlen` should still be in the corresponding register from the `gnrc_pktbuf_add()` call.

miri64 commented on drivers/include/cc1xxx_common.h:
```diff
@@ -0,0 +1,138 @@
+ * | Source      | The layer 2 source address      |
+ * | Payload     | The payload (variable size)     |
+ *
+ * Please note that the payload needs to be able to carry 6LoWPAN frames.
```
Again this is too specific to IPv6-over-cc1xxx for my taste. If people want to use something else this is also possible.

miri64 commented on drivers/include/cc1xxx_common.h:
```diff
@@ -0,0 +1,138 @@
+ * This adaption layer assumes that the layer 2 address `0x00` is reserved for
+ * layer 2 broadcast, which is true for CC110x and CC1200 transceivers (provided
+ * they are configured accordingly). If more users of this adaption layer are
+ * added, this behaviour might needs to be more generalized.
```
I would add some reference to `@ref CC1XXX_BCAST_ADDR` somewhere in this paragraph, but that is not a must.

maribu commented:
@miri64: Thank you for your review. I think I addressed everything and squashed. The unsquashed version backed up [here](https://github.com/maribu/RIOT/tree/gnrc_netif_cc1xxx_unsquashed) for reference.

miri64 commented on drivers/include/cc1xxx_common.h:
```diff
@@ -0,0 +1,137 @@
+ * more users of this adaption layer are added, this behaviour might needs to be
+ * more generalized.
+ */
+#ifndef NET_GNRC_NETIF_CC1XXX_H
```
Needs to be changed as pointed out by both Travis and Murdock ;-)

maribu commented:
Fixed and squashed. (See https://github.com/maribu/RIOT/tree/gnrc_netif_cc1xxx_unsquashed for the unsquashed version of this PR)

miri64 commented:
ACK upholds.

miri64 commented:
And go
